src/kernel/filesystem/ram_device.zig

```diff
@@ -0,0 +1,228 @@
+const std = @import("std");
+const expectEqual = std.testing.expectEqual;
+const expectError = std.testing.expectError;
+const expectEqualSlices = std.testing.expectEqualSlices;
+const Device = @import("device.zig").Device;
+const Allocator = std.mem.Allocator;
+
+/// A simple RAM device. This can be use for easy testing of file systems. This can also be use for
+/// easy creation of in memory file systems.
+pub const RamDevice = struct {
+    const Self = @This();
+
+    /// The underlying device interface.
+    device: *Device,
+
+    /// An allocator to allocate memory when needed.
+    allocator: *Allocator,
+
+    /// The memory that the RAM device will control.
+    memory: []u8,
+
+    /// See Device.instance.
+    instance: usize,
+
+    /// See Device.read
+    fn read(device: *const Device, offset: usize, len: usize) Device.Error![]u8 {
+        var self = @fieldParentPtr(RamDevice, "instance", device.instance);
+        if (self.memory.len - len < offset) {
+            return Device.Error.OutOfBounds;
+        }
+        var buff = try self.allocator.alloc(u8, len);
+        std.mem.copy(u8, buff[0..], self.memory[offset .. offset + len]);
+        return buff;
+    }
+
+    /// See Device.write
+    fn write(device: *const Device, offset: usize, bytes: []const u8) Device.Error!void {
+        var self = @fieldParentPtr(RamDevice, "instance", device.instance);
+        if (self.memory.len - bytes.len < offset) {
+            return Device.Error.OutOfBounds;
+        }
+        std.mem.copy(u8, self.memory[offset..], bytes[0..]);
+    }
+
+    ///
+    /// De-initialise a RAM device. This will free the underlying device, the memory this device
+    /// controls and this device itself.
+    ///
+    /// Argument:
+    ///     IN self: *Self - Self.
+    ///
+    pub fn deinit(self: *Self) void {
+        self.allocator.destroy(self.device);
+        self.allocator.free(self.memory);
+        self.allocator.destroy(self);
+    }
+
+    ///
+    /// Initialise a new RAM device with the memory provided duplicated. After this is called, the
+    /// memory provided can be freed.
+    ///
+    /// Arguments:
+    ///     IN memory: []u8          - The pointer to the bytes that the RAM device will manage.
+    ///     IN allocator: *Allocator - The allocator when needing to allocate memory.
+    ///
+    /// Return: *RamDevice
+    ///     A new RAM device.
+    ///
+    /// Error: Allocator.Error
+    ///     error.OutOfMemory - If there isn't enough memory when initialising the RAM device.
+    ///                         Any memory allocated will be freed on return.
+    ///
+    pub fn init(memory: []u8, allocator: *Allocator) Allocator.Error!*RamDevice {
```

Comment:
Could this return a struct instead to allow the caller to decide if it should be on the heap or stack?

Reply:
Sure